controlplane/pkg/nsm/nsm_heal.go

```diff
@@ -10,32 +10,31 @@ import (
 
 func (srv *networkServiceManager) Heal(connection nsm.NSMClientConnection, healState nsm.HealState) {
 	healID := create_logid()
-	logrus.Infof("NSM_Heal(1-%v) %v", healID, connection)
+	logrus.Infof("NSM_Heal(%v) %v", healID, connection)
 
 	cc := connection.(*model.ClientConnection)
-	if cc.ConnectionState != model.ClientConnectionReady {
-		//means that we already closing/healing
+	cc.ConnectionState = model.ClientConnectionHealing
+	if ok := srv.model.CompareAndSwapClientConnection(cc, func(connection *model.ClientConnection) bool {
+		return connection.ConnectionState == model.ClientConnectionReady
```

Comment:
Hm, it's a bit strange, for example, `ClientConnection.Xcon` could be changed in the model, but we will overwrite it with `cc.Xcon` which is outdated

Reply:
This has been completely reworked with `ClientConnectionEditor`.